src/scale.jl

```diff
@@ -251,45 +253,52 @@ function apply_scale_typed!(ds, field, scale::ContinuousScale)
-    level_values = levels(da)
+function reorder_levels(da::IndirectArray, order::AbstractVector)
+    level_values = da.values
     length(order) != length(level_values) &&
             error("Discrete scale order is not of the same length as the data's levels.")
     permute!(level_values, order)
-    return PooledDataArray(da, level_values)
+    return IndirectArray(da, level_values)
 end
 
-function discretize_make_pda(values::Vector, levels=nothing)
-    if levels == nothing
-        return PooledDataArray(values)
-    else
-        return PooledDataArray(convert(Vector{eltype(levels)}, values), levels)
-    end
-end
+discretize_make_ia(values::Vector)         = IndirectArray(values)
+discretize_make_ia(values::Vector, ::Void) = IndirectArray(values)
+discretize_make_ia(values::Vector, levels) =
+    IndirectArray(map(t -> findfirst(levels, t), values), levels)
 
-function discretize_make_pda(values::DataArray, levels=nothing)
-    if levels == nothing
-        return PooledDataArray(values)
-    else
-        return PooledDataArray(convert(DataArray{eltype(levels)}, values), levels)
-    end
-end
+discretize_make_ia(values::DataArray)         = IndirectArray(values)
+discretize_make_ia(values::DataArray, ::Void) = IndirectArray(values)
+discretize_make_ia(values::DataArray, levels) =
+    IndirectArray(convert(DataArray{eltype(levels)}, values), levels)
 
-function discretize_make_pda(values::Range, levels=nothing)
-    if levels == nothing
-        return PooledDataArray(collect(values))
-    else
-        return PooledDataArray(collect(values), levels)
-    end
-end
+discretize_make_ia(values::Range)         = IndirectArray(collect(values))
+discretize_make_ia(values::Range, ::Void) = IndirectArray(collect(values))
+discretize_make_ia(values::Range, levels) = IndirectArray(collect(values), levels)
 
-function discretize_make_pda(values::PooledDataArray, levels=nothing)
-    if levels == nothing
-        return values
-    else
-        return PooledDataArray(values, convert(Vector{eltype(values)}, levels))
-    end
+discretize_make_ia(values::IndirectArray)         = values
+discretize_make_ia(values::IndirectArray, ::Void) = values
+discretize_make_ia(values::IndirectArray, levels) = IndirectArray(values, levels)
+
+discretize_make_ia(values::CategoricalArray) =
+    discretize_make_ia(values, intersect(push!(levels(values), missing), unique(values)))
+discretize_make_ia(values::CategoricalArray, ::Void) = discretize_make_ia(values)
+function discretize_make_ia(values::CategoricalArray{T}, levels::Vector) where {T}
+    index = map!(t -> ismissing(t) ?
+                    findfirst(ismissing, levels) :
+                    findfirst(s -> !(ismissing(s) || s != get(t)), levels),
+                 Array{UInt8}(size(values)...),
+                 values)
+    return IndirectArray(index, convert(Vector{T},levels))
 end
+function discretize_make_ia(values::CategoricalArray{T}, levels::CategoricalVector{T}) where T
```

Comment:
Really there should be no need for this special method. `::AbstractArray` should work equally for this one and the one above.

Reply:
Most of the time spent (which now was quite a bit) was spent on making sure that no wrapped types made it through to `Compose`. The problem with `AbstractArray` is that you don't know if the elements are wrapped on not. Concrete methods handle the cases explicitly. We might be able to clean this up but these are internal methods so that can be done later if anybody feels for doing it.